src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -701,12 +705,13 @@ private Stream OpenInWriteMode()
                                                             (object o, EventArgs e) =>
                                                             {
                                                                 //release the archive stream
-                                                                _archive.ReleaseArchiveStream(this);
-                                                                _outstandingWriteStream = null;
+                                                                var entry = (ZipArchiveEntry)o;
```

Comment:
GitHub won't let me leave a comment on L735, but could you do a similar cast-to-avoid-closure in the lambda in OpenInUpdateMode?